lib/iolog/log_writer.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iolog
+
+import (
+	"bytes"
+	"encoding/binary"
+	"io"
+	"time"
+)
+
+// Clock defines an interface that wraps time.Now()
+type Clock interface {
+	Now() time.Time
+}
+
+// LogClock is an implementation of the Clock interface that
+// returns time.Now() for use in the logio package
+type LogClock struct{}
+
+// Now returns the local time time.Now()
+func (LogClock) Now() time.Time {
+	return time.Now().UTC()
+}
+
+// LogWriter tags log entries with a descriptive header and writes
+// them to the underlying io.Writer
+type LogWriter struct {
+	Clock
+	w io.Writer
+}
+
+// NewLogWriter wraps an io.WriteCloser in a LogWriter
+func NewLogWriter(w io.Writer, clock Clock) *LogWriter {
+	return &LogWriter{
+		w:     w,
+		Clock: clock,
+	}
+}
+
+// Write scans the supplied buffer, breaking off indvidual entries
+// and writing them to the underlying Writer, flushing any leftover bytes
+func (lw *LogWriter) Write(p []byte) (n int, err error) {
+	var (
+		start, end, i int
+		entry         []byte
+	)
+	// TODO(jzt): traverse the buffer byte-by-byte and write a chunk
+	// when we hit either a '\n' or maxEntrySizeBytes, to improve readibility
+	for {
+		i = 0
+		if i = bytes.IndexByte(p[start:], '\n') + 1; i == 0 {
+			break
+		}
+		end = start + i
+		entry = p[start:end]
+		// we have a complete entry, let's write it
+		n, err = lw.write(entry)
+		if err != nil {
+			return n, err
+		}
+		// advance starting index
+		start = end
+	}
+	// flush the buffer if needed
+	if start < len(p) {
+		lw.write(p[start:])
```

Comment:
The header adds 10 bytes, but otherwise the entry stays exactly the same. Most writes are likely fairly small (human readable text in reasonably-sized entries, if I had to guess).